modules/core/src/main/java/org/apache/ignite/metric/IgniteMetricRegistry.java

```diff
@@ -0,0 +1,103 @@
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.apache.ignite.lang.IgniteExperimental;
+import org.apache.ignite.spi.metric.BooleanMetric;
+import org.apache.ignite.spi.metric.DoubleMetric;
+import org.apache.ignite.spi.metric.IntMetric;
+import org.apache.ignite.spi.metric.LongMetric;
+import org.apache.ignite.spi.metric.ObjectMetric;
+import org.apache.ignite.spi.metric.ReadOnlyMetricRegistry;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Metric registry. Allows to get, add or remove metrics.
+ *
+ * @see IgniteMetrics
+ * @see ReadOnlyMetricRegistry
+ */
+@IgniteExperimental
+public interface IgniteMetricRegistry extends ReadOnlyMetricRegistry {
+    /**
+     * Registers an int metric which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     * @return New or previously registered {@link IntMetric}.
+     */
+    IntMetric register(String name, IntSupplier supplier, @Nullable String desc);
+
+    /**
+     * Registers a long which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     * @return New or previously registered {@link LongMetric}.
+     */
+    LongMetric register(String name, LongSupplier supplier, @Nullable String desc);
+
+    /**
+     * Registers a double metric which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     * @return New or previously registered {@link DoubleMetric}.
+     */
+    DoubleMetric register(String name, DoubleSupplier supplier, @Nullable String desc);
+
+    /**
+     * Registers an object metric which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param type Metric value type.
+     * @param desc Metric description.
+     * @param <T> Metric value type.
+     * @return New or previously registered {@link ObjectMetric}.
+     */
+    <T> ObjectMetric<T> register(String name, Supplier<T> supplier, Class<T> type, @Nullable String desc);
+
+    /**
+     * Registers a boolean metric which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     * @return New or previously registered {@link BooleanMetric}.
+     */
+    BooleanMetric register(String name, BooleanSupplier supplier, @Nullable String desc);
+
+    /**
+     * Removes metrics with the {@code name}.
+     *
+     * @param name Metric name..
+     */
+    void remove(String name);
+
+    /** Resets all metrics of this metric registry. */
+    void reset();
```

Comment:
Can you, please, clarify what "reset metric" means exactly?
All metrics registered with public API have suppliers to provide current value of the metric.
It's not clear what this method will do.

Reply:
Changed the javadoc to 'Calls {@link Metric#reset()} for all the registered metrics.'. Is that enough?

Reply:
Why do we need this method, in the first place?

Reply:
What do you mean 'in the first place'? Dropping, resetting, of a user metric looks intrinsic. Why not?

Reply:
Because, all metrics we can create with the current API are those that works with `Supplier`.
And `reset` method intention is to assign 0 or some default value to the counter or something similar.
Please, take a look at `IntGauge` and other similar implementions.

Reply:
Yes, but we provide `reset()` even for the gauges.

Reply:
`reset` for gauges are no-op.